strutil/set.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// OrderedSet is a set of strings that maintains the order of insertion.
+//
+// External synchronization is required for safe concurrent access.
+type OrderedSet struct {
+	items      []string
+	positionOf map[string]int
```

Comment:
Just an observation (in case memory was a concern here, but probably isn't), we could avoid storing items and just reconstruct them when needed from positionOf.

Reply:
Yeah that is true. I had a TODO that I removed from the implementation of `Items()` that we can essentially look at the map and craft the order.

The complexity of the current `Items()` method is O(N), if we drop the `items` field then items would roughly look like:

```
items := make([]string, len(os.positionOf)
for item, idx := range positionOf {
    items[idx] = item
}
return items
```

Which is also O(N) but at O(N) less memory than originally in the stationary case. I can make the change if you feel like it. I was mainly going for simple-and-correct since the original problem was more interesting but I don't mind optimising this slightly, especially since it seems it would not be hard to do so.

Reply:
No strong opinion, as I said it's more of an observation. Perhaps just add a comment next to items member saying it's for convienience and can be inferred from positionOf.

Reply:
I'll just implement it :)

Reply:
Done, marking as resolved.